### PR TITLE
Allow recipes to inherit from defaults

### LINKED ISSUE #116: Recipe inheritance
Inherit recipe keywords from the formula in the recipe repositories by default; provide user option and keyword to disable.

### PR DESCRIPTION
**Pull request by CeleritasCelery**
`:host`, `:repo`, and `:files` are inherited from the default recipe. Also `:fork` will inherit its host from the upstream recipe. These can be overridden like any recipe component.

closes #116 

For example, we can do the following simplification
### example 1
before:
```lisp
(package :host gitlab :repo "other-user/repo" 
         :files ("lisp.el")
         :fork (:host :gitlab :repo "my-user/repo"))
```
after:
```lisp
(package :fork "my-user/repo")
```
or 
```lisp
(package :fork (:repo "my-user/repo"))
```
### example 2
before:
```lisp
(package :host gitlab :repo "user/repo" :local-repo "local")
```

after:
```lisp
(package :local-repo "local")
```

Let me know what you think. If you give the green light I will add some documentation to the main README.

### CODE REVIEW

**Comment by vyp** [comment]
What about when someone has `straight-allow-recipe-inheritance` set to `t`, but doesn't want say `:files` to be inherited in one particular recipe? (Assuming that when a recipe doesn't have `:files`, straight uses some default `:files` value?)

In that spirit, have you considered maybe having some way to enable/disable inheritance in a recipe itself, like say `:inherit` keyword?

**Comment by CeleritasCelery** [comment]
> What about when someone has straight-allow-recipe-inheritance set to t, but doesn't want say :files to be inherited in one particular recipe?

All of these properties can be overridden. So if someone does not want `:files` to be inherited they can set it explicitly in their recipe. 

The whole idea behind this PR is that you should only need to override the properties you actually care about and not need to add redundant information that straight already knows. This does not prevent you from completely overriding a recipe If you so choose.

I feel like an `:inherit` keyword is unnecessary because this is the sensible default behavior and if you don’t like it, you can override the all recipes components explicitly (which is what you have to do now). I considered not even adding a variable to control this because all `straight-allow-recipe-inheritance` does it requires you to be explicit about everything (leading to a longer more error prone configuration).

**Comment by vyp** [comment]
Yeah I understand the reason for this PR, I've been wanting this feature for a while too. :tada: Anyway I found out that I can just specify `:files :defaults` to set the default value for `:files`, so yeah anything like `:override` wouldn't be needed. :+1:

**Comment by raxod502** [comment]
This sounds really lovely, it's a feature that has been sorely lacking for quite some time now. I will get to the review as soon as I can.

File: straight.el [hunk 1]
```diff
@@ -208,6 +208,27 @@ profile (nil) will suffice without additional setup."
...
+
+can be simplifed to
+
+'(package :fork \"my-user/repo\")
+
+if the original recipe was
+
+'(package :host gitlab :repo \"other-user/repo\")"
+  :type 'boolean)
+
```
Comment (raxod502) [inline straight.el]: An important shortcoming should be documented: namely, that if recipe inheritance is enabled then all recipe repositories need to be cloned right away even if you're specifying a custom recipe for a package that doesn't exist in any of them.

File: straight.el [hunk 2]
```diff
@@ -208,6 +208,27 @@ profile (nil) will suffice without additional setup."
                 (alist :key-type symbol :value-type
                        (plist :key-type symbol :value-type sexp))))
 
+(defcustom straight-allow-recipe-inheritance t
+  "Non-nil allows missing recipe components to be inherited.
+These missing components inherit from the default recipe. The
+supported components are `:host', `:repo', `:files', and `:fork'.
+`:fork' will inherit the `:host' component from the original
+recipe origin.
```
Comment (raxod502) [inline straight.el]: I would like significantly more explicitness about exactly what this means. For example, it's not clear at all from this description that if you specify a string for `:fork` it gets translated into a list `(:repo "...")` and then additional attributes are copied over if non-nil (is that the behavior? whatever the behavior is should be made explicit and defended).

File: straight.el [hunk 3]
```diff
@@ -208,6 +208,27 @@ profile (nil) will suffice without additional setup."
...
+
+For example, the following recipe
+
+'(package :host gitlab :repo \"other-user/repo\" :fork (:host
+:gitlab :repo \"my-user/repo\"))
```
Comment (raxod502) [inline straight.el]: Avoid curly quotes: https://github.com/raxod502/straight.el/blob/6649858bdb76d0c557bf46c93b3a2eb539cec750/straight.el#L1209

Having this indented nicely would also be desirable, I think.

File: straight.el [hunk 4]
```diff
@@ -2866,6 +2887,27 @@ for dependency resolution."
           ;; override the default value (which is determined according
           ;; to the selected VC backend).
           ;;
+          (when straight-allow-recipe-inheritance
+            ;; To keep overridden recipes simple, the `:host',
+            ;; `:repo', `:files' can be inherited from the original
+            ;; recipe. Also `:fork' can inherit its host from origin,
```
Comment (raxod502) [inline straight.el]: Does "from origin" mean "from the original recipe"?

File: straight.el [hunk 5]
```diff
@@ -2866,6 +2887,27 @@ for dependency resolution."
...
+            ;; meaning that :fork "user/repo" is equivalent to :fork
+            ;; (:repo "user/repo")
+            (when-let ((fork (plist-get plist :fork))
+                       (_stringp (stringp fork)))
```
Comment (raxod502) [inline straight.el]: I would prefer just `let` with `when` inside.

File: straight.el [hunk 6]
```diff
@@ -2866,6 +2887,27 @@ for dependency resolution."
...
+                (cdr (straight-recipes-retrieve package))
+              (when (and host (not (plist-member plist :host)))
+                (straight--put plist :host host))
+              (when (and repo (not (plist-member plist :repo)))
+                (straight--put plist :repo repo))
+              (when (and files (not (plist-member plist :files)))
+                (straight--put plist :files files))
+              (when-let ((host host)
+                         (fork-plist (plist-get plist :fork))
+                         (_no-fork-host (not (plist-member fork-plist :host))))
```
Comment (raxod502) [inline straight.el]: Large `when-let` seems a bit hard to follow to me. Can we just split it into multiple nested conditions which are simpler?

File: straight.el (hunk 1 above)
Comment (CeleritasCelery) [reply straight.el]: Expand on this. What would a user have to do to make sure recipe repos were cloned at the start?

File: straight.el (hunk 3 above)
Comment (CeleritasCelery) [reply straight.el]: Not sure what you mean by "avoid curly quotes".  Are you saying that I should be using single quotes in the documentation even though that is invalid syntax?

Comment (CeleritasCelery) [reply straight.el]: nevermind, I see what you mean.

File: straight.el (hunk 1 above)
Comment (raxod502) [reply straight.el]: I don't mean they need to be cloned by the user. I mean that enabling recipe inheritance will cause `straight.el` to automatically clone them all. This isn't otherwise necessary if you provide an explicit recipe (as no recipe needs to be looked up).

I do not think this is a big problem, but I think it is worth mentioning.

Comment (CeleritasCelery) [reply straight.el]: How is this triggered? With `straight-recipes-retrieve`?

**Comment by CeleritasCelery** [comment]
I addressed your feedback. Please take another look.

File: straight.el (hunk 1 above)
Comment (raxod502) [reply straight.el]: Yes.

File: straight.el [hunk 7]
```diff
@@ -2912,6 +2942,24 @@ for dependency resolution."
...
+          (when straight-allow-recipe-inheritance
+            ;; To keep overridden recipes simple, the
+            ;; `straight-vc-git-keywords' and `:files' can be
+            ;; inherited from the original recipe. Also `:fork' can
+            ;; inherit its host from the original recipe.
+            (let ((fork (plist-get plist :fork)))
+              (when (stringp fork)
+                (straight--put plist :fork `(:repo ,fork))))
+            (let ((default (cdr (straight-recipes-retrieve package))))
+              (dolist (keyword (cons :files (straight-vc-git-keywords)))
```
Comment (raxod502) [inline straight.el]: Don't use `straight-vc-git-keywords`, instead use `straight-vc-keywords` with the appropriate `:type` for this recipe.

File: straight.el [hunk 8]
```diff
@@ -2912,6 +2942,24 @@ for dependency resolution."
           ;; override the default value (which is determined according
           ;; to the selected VC backend).
           ;;
+          (when straight-allow-recipe-inheritance
+            ;; To keep overridden recipes simple, the
+            ;; `straight-vc-git-keywords' and `:files' can be
+            ;; inherited from the original recipe. Also `:fork' can
+            ;; inherit its host from the original recipe.
```
Comment (raxod502) [inline straight.el]: This comment is out of date.

File: straight.el [hunk 9]
```diff
@@ -208,6 +208,36 @@ profile (nil) will suffice without additional setup."
...
+  "Non-nil allows partially overriding recipes.
+If you override a recipe, every component that is not explicitly
+overriden will be searched for in original recipe. If found, that
+value will be added to the overriden recipe. This allows you to
+only override the recipe components you are interested in,
+instead of being required to override them all. The supported
+components are the ones listed by `straight-vc-git-keywords' and
+`:files'. Note that enabling this feature has the side effect
+that all recipe repos (i.e. melpa, elpa) will always be cloned,
+even if none of the recipes are used.
```
Comment (raxod502) [inline straight.el]: I would say, "even if you explicitly specify all your recipes".

File: straight.el [hunk 10]
```diff
@@ -208,6 +208,36 @@ profile (nil) will suffice without additional setup."
...
+
+The `:fork' keyword is handled specially. If its value is a
+string instead of a list, then it is assigned as the `:repo' of
+the fork. Also the fork recipe will inherit its `:host' component
+from the default recipe.
```
Comment (raxod502) [inline straight.el]: Why should `:fork` only inherit `:host`, rather than all of `:repo`, `:branch`, and `:remote` as well?

**Comment by CeleritasCelery** [comment]
Please review.

**Comment by raxod502** [comment]
Thanks for the contribution!

**Comment by CeleritasCelery** [comment]
Happy to contribute to one of my favorite packages.
